lib/membrane_wav/parser.ex

```diff
@@ -0,0 +1,243 @@
+      the header is correct, but does not use that data in any way. After parsing, the stage is
+      set to `:data`.
+    - `:data` - header is already fully parsed. All new data from the input is sent to the output.
+  """
+
+  use Membrane.Filter
+
+  alias Membrane.Buffer
+  alias Membrane.Caps.Audio.Raw, as: Caps
+  alias Membrane.Caps.Audio.Raw.Format
+
+  require Membrane.Logger
+
+  @pcm_format_size 16
+
+  @init_stage_size 22
+  @format_stage_size 22
+  @fact_stage_base_size 8
+
+  def_options frames_per_buffer: [
+                type: :integer,
+                spec: pos_integer(),
+                description: """
+                Assumed number of raw audio frames in each buffer.
+                Used when converting demand from buffers into bytes.
+                """,
+                default: 2048
+              ]
+
+  def_output_pad :output,
+    mode: :pull,
+    availability: :always,
+    caps: Caps
+
+  def_input_pad :input,
+    mode: :pull,
+    availability: :always,
+    demand_unit: :bytes,
+    caps: Caps
+
+  @impl true
+  def handle_init(state) do
+    state = Map.merge(state, %{stage: :init, queue: <<>>})
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_start_of_stream(:input, _context, state) do
+    demand = {:input, @init_stage_size}
+
+    {{:ok, demand: demand}, state}
+  end
+
+  @impl true
+  def handle_demand(:output, size, :bytes, _context, %{stage: :data} = state) do
+    {{:ok, demand: {:input, size}}, state}
+  end
+
+  def handle_demand(
+        :output,
+        buffers_count,
+        :buffers,
+        _context,
+        %{stage: :data, frames_per_buffer: frames, caps: caps} = state
+      ) do
+    size = buffers_count * Caps.frames_to_bytes(frames, caps)
+
+    {{:ok, demand: {:input, size}}, state}
+  end
+
+  def handle_demand(:output, _size, _unit, _context, state) do
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_event(_pad, event, _context, state) do
+    Membrane.Logger.debug("Received event #{inspect(event)}")
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_process(:input, buffer, context, %{stage: :data} = state) do
+    if context.pads.input.demand == 0 do
+      {{:ok, buffer: {:output, buffer}, redemand: :output}, state}
+    else
+      {{:ok, buffer: {:output, buffer}}, state}
+    end
+  end
+
+  def handle_process(:input, %Buffer{payload: payload} = _buffer, context, state)
+      when byte_size(payload) < context.pads.input.demand do
+    demand_fun = &(&1 - byte_size(payload))
+    state = Map.update(state, :queue, payload, &(&1 <> payload))
+
+    {{:ok, demand: {:input, demand_fun}}, state}
+  end
+
+  def handle_process(
+        :input,
+        %Buffer{payload: payload} = _buffer,
+        _context,
+        %{stage: :init, queue: queue} = state
+      ) do
+    "RIFF" <>
+      <<_file_size::binary-size(4)>> <>
+      "WAVE" <>
+      "fmt " <>
+      <<format_size::binary-size(4)>> <>
+      <<format::binary-size(2)>> = queue <> payload
```

Comment:
You may patter-match without using concatenation operator and convert to integer right away: 
```
<<"RIFF", size :: 32, "WAVE" >> = "RIFF" <> <<0, 0, 0, 1>> <> "WAVE"
```
so you can do (requires formatting):
```suggestion
    <<"RIFF", _file_size::32, "WAVE", "fmt ", format_size::signed-little-size(4)-unit(8), format::signed-little-size(2)-unit(8)>> = queue <> payload
```

Reply:
That's good. It looked ugly with all that concatenations.